### PR TITLE
[serve] Add unimplemented interfaces for Deployment DAG APIs

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Adds the following interfaces (without implementation, for discussion / approval):
- `serve.Application`
- `serve.DeploymentNode`
- `serve.DeploymentMethodNode`, `serve.DAGHandle`, and `serve.drivers.PipelineDriver`
- `serve.run` & `serve.build`

In addition to these Python APIs, we also support the following CLI commands:
- `serve run [--blocking=true] my_file:my_node_or_app # Uses Ray client, blocking by default.`
- `serve build my_file:my_node output_path.yaml`
- `serve deploy [--blocking=false] # Uses REST API, non-blocking by default.`
- `serve status [--watch=false] # Uses REST API, non-blocking by default.`

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/api.py [hunk 1]
```diff
@@ -1451,3 +1508,203 @@ def get_deployment_statuses() -> Dict[str, DeploymentStatusInfo]:
...
+    """
+    raise NotImplementedError()
+
+
+@PublicAPI(stability="alpha")
+def build(
+    target: BoundDeploymentNode,
+    *,
+    host: str = DEFAULT_HTTP_HOST,
+    port: int = DEFAULT_HTTP_PORT,
```
Comment (edoakes) [inline python/ray/serve/api.py:1692]: maybe use HTTPOptions ?

**Comment by edoakes** [comment]
@simon-mo I would prefer we don't use a nested "Options" for .run to match other HTTP serving libraries (e.g., `uvicorn.run("example:app", host="127.0.0.1", port=5000, log_level="info")`

I also don't think there should be a significant number of new options here

File: python/ray/serve/api.py [hunk 2]
```diff
@@ -890,6 +894,57 @@ async def __del__(self):
     return decorator
 
 
+@PublicAPI(stability="alpha")
+class DAGHandle:
+    """Resolved from a DeploymentMethodNode at runtime.
```
Comment (jiaodong) [inline python/ray/serve/api.py:899]: To be accurate i think we should say this is the DAGNode of user chose acting as the root, where all DAGNodes referenced becomes children. It doesn't have to be deployment or deployment method node in particular. We support using a plain ray task as well.

File: python/ray/serve/api.py [hunk 3]
```diff
@@ -890,6 +894,57 @@ async def __del__(self):
...
+    def __init__(self, serialized_dag_json: str):
+        raise NotImplementedError()
+
+    def remote(self, *args, **kwargs) -> ray.ObjectRef:
+        """Call the DAG."""
+        raise NotImplementedError()
+
+
+@PublicAPI(stability="alpha")
+class DeploymentMethodNode(ClassNode):
```
Comment (jiaodong) [inline python/ray/serve/api.py:914]: this subclasses DAGNode. I think we can just import the existing one ?

File: python/ray/serve/api.py [hunk 4]
```diff
@@ -890,6 +894,57 @@ async def __del__(self):
...
+    This class cannot be called directly. Instead, when it is bound to a
+    deployment node, it will be resolved to a DeployedCallGraph at runtime.
+    """
+
+    def __init__(self):
+        raise NotImplementedError()
+
+
+@PublicAPI(stability="alpha")
+class DeploymentNode(ClassNode):
```
Comment (jiaodong) [inline python/ray/serve/api.py:929]: ditto

File: python/ray/serve/api.py (hunk 3 above)
Comment (edoakes) [reply python/ray/serve/api.py:914]: IMO the existing one should be moved here w/ other top-level Serve APIs.

File: python/ray/serve/api.py [hunk 5]
```diff
@@ -1451,3 +1508,180 @@ def get_deployment_statuses() -> Dict[str, DeploymentStatusInfo]:
...
+    def __contains__(self, key: str):
+        """Checks if the key exists in self._deployments."""
+        raise NotImplementedError()
+
+
+@PublicAPI(stability="alpha")
+def run(
+    target: Union[DeploymentNode, Application],
+    *,
+    blocking: bool = False,
```
Comment (edoakes) [inline python/ray/serve/api.py:1651]: we could consider removing this and only supporting it via CLI

File: python/ray/serve/api.py (hunk 3 above)
Comment (jiaodong) [reply python/ray/serve/api.py:914]: im good with that and it helps to resolve circular imports

Comment (edoakes) [reply python/ray/serve/api.py:914]: yup. let's plan to do that.

**Comment by edoakes** [comment]
@simon-mo re: ingress_kwargs, I think host and port are generic to any ingress type so they're ok being actual kwargs, but if we add more in the future then yes.

**Comment by simon-mo** [comment]
sgtm

File: python/ray/serve/api.py (hunk 3 above)
Comment (jiaodong) [inline python/ray/serve/api.py:914]: subclass `DAGNode` is better here, even current `ClassMethodNode` don't use much of `ClassNode` implementation

later we likely also convert plain task / function to `DeploymentNode` that wraps a func_or_class, so better not to assume it's always an instance of ClassNode.

File: python/ray/serve/api.py [hunk 6]
```diff
@@ -1451,3 +1508,157 @@ def get_deployment_statuses() -> Dict[str, DeploymentStatusInfo]:
     """
 
     return internal_get_global_client().get_deployment_statuses()
+
+
+class ImmutableDeploymentDict(dict):
+    def __init__(self, deployments: List[Deployment]):
```
Comment (shrekris-anyscale) [inline python/ray/serve/api.py:1514]: Should this take in a dictionary or a list?

File: python/ray/serve/api.py (hunk 3 above)
Comment (edoakes) [reply python/ray/serve/api.py]: done

File: python/ray/serve/drivers.py [hunk 7]
```diff
@@ -0,0 +1,21 @@
+from typing import Callable, Optional, Union
+
+import starlette
+
+from ray import serve
+from ray.serve.api import DeployedCallGraph
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/drivers.py:6]: oops

[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/drivers.py:6]: should be DAGHandle, fixed

File: python/ray/serve/api.py (hunk 6 above)
Comment (edoakes) [reply python/ray/serve/api.py:1514]: this is an internal API so whichever you find easier in impl

File: python/ray/serve/drivers.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
+from typing import Callable, Optional, Union
+
+import starlette
+
+from ray import serve
+from ray.serve.api import DAGHandle
+
+
+@serve.deployment
+class PipelineDriverDeployment:
```
Comment (simon-mo) [inline python/ray/serve/drivers.py:10]: non blocking, should this be brought to top level import? and call it DefaultDriver or something?

**Review by simon-mo — APPROVED** [review]

File: python/ray/serve/drivers.py (hunk 8 above)
Comment (edoakes) [reply python/ray/serve/drivers.py:10]: possibly, I'm thinking that we will probably end up with a few of these (i.e., a gRPC one, one that does HTTP routing, etc.) so having one submodule where they're all defined might make sense

Comment (edoakes) [reply python/ray/serve/drivers.py]: I renamed it to `from.serve.ingress import DAGDriver` for consistency with other names

**Pull request merged**